### ISSUE TITLE
Cropping Issue in HE

### ISSUE THREAD

**Issue by AureoZ** [Teichlab/bin2cell#22]
Hi,
Great work on bin2cell the capabilities of the software are truly outstanding.

I am currently running "b2c.scaled_he_image()" however the subsequent "he.tiff", results in the background image of my visualizations being cropped in a region that doesn't correspond to the 008um data, however when I have utilized the cyt-assist image the overlay works without issue and the corresponding areas on the mask align with the background. I've attempted to turn off the cropping via "crop=False", however it seems too memory intensive and returns me "OutOfMemoryError".  The image I am utilizing is the image used in the space ranger argument for "--image" and is roughly 4 GB in size.

Is there any way for me to fix this without just simply throwing more computational power at it?

**Comment by nadavyayon** [comment]
hi, is there by chance of a mismatch between your image and the expression data? did you have a look in loupe browser? For me sometimes the alignment between the CytAssist image and the microscope image is way off to the extent you don't see your image. If this is the case and in any case i would recommend performing manual alignment via their (10X) dedicated tool in Loupe

**Comment by ktpolanski** [comment]
I don't think it's that, Nadav. Otherwise they wouldn't be saying the 8um data looks different.

That said, I don't understand what you're trying to convey. The CytAssist image is not utilised in any form by b2c, and inserting it as an in-place replacement for the morphology image won't get you anywhere as the spatial coordinates mean nothing there.

Please visualise the full morphology image. Please visualise the `hires` image from spaceranger's output. Please visualise the cropped `he.tiff` image.

**Comment by AureoZ** [comment]
Here is a visualization of the images and also a visualization of the current issue I am having. Thank you for all your help thus far!

![trb](https://github.com/user-attachments/assets/a51777e9-a1ec-4f77-8551-ddde4997c2e0)

![test](https://github.com/user-attachments/assets/95cc17d9-5ce5-4508-a3c3-3765764ae1f2)

**Comment by ktpolanski** [comment]
Thanks. This doesn't make much sense to me - `he.tiff` looks like a whole different image, and not a particularly useful one. I'm having a hard time believing this is a bug, as you'd think something so glaring would have come up across however many samples were analysed with this before.

Does `sc.pl.spatial(adata, color=[None, "n_counts", "n_counts_adjusted"])`, without specifying the custom H&E image via extra parameters, look accurate? I expect it to, just double checking.

What happens when you try to load and visualise the `source_image_path` as passed to `b2c.read_visium()`?
```python
import matplotlib.pyplot as plt

img = b2c.load_image(source_image_path)
img.shape
plt.imshow(img)
```

**Comment by AureoZ** [comment]
When visualizing the image via the provided code, the result is the following,
![trb2](https://github.com/user-attachments/assets/a8cf82fa-6282-4f64-be13-35b84af5ad9e)

Using "sc.pl.spatial(adata, color=[None, "n_counts", "n_counts_adjusted"])" results in this 
![test0](https://github.com/user-attachments/assets/c6fa434a-5b5d-4a73-8ca3-0d747da76039)

**Comment by ktpolanski** [comment]
The other plot doesn't look great either. It's strange that somehow the two disagree, I've never seen this sort of thing before.

Could I get access to these files? Think it would be easier for me to try to experiment on them myself, especially as I don't currently have a hypothesis as to what is going on.

**Comment by ktpolanski** [comment]
Meanwhile I came up with a couple more diagnostics that can be done on your end:
- Please report what `np.min(adata.obsm['spatial'], axis=0)` and `np.max(adata.obsm['spatial'], axis=0)` return. Please make sure that `adata` is not a crop, but rather the whole loaded object.
- Please share the contents of `scalefactors_json.json`, which can be found in `binned_outputs/square_002um/spatial`

**Comment by AureoZ** [comment]
For np.min(adata.obsm['spatial'], axis=0) the result is,
"array([378.10206736, 745.19454447])"

For "np.max(adata.obsm['spatial'], axis=0)" the result is, 
"array([1811.25910355, ])"

The contents of the "scalefactors_json.json" file are as follows,

{
    "spot_diameter_fullres": 0.4277836479625891,
    "bin_size_um": 2.0,
    "microns_per_pixel": 4.675260519015691,
    "regist_target_img_scalef": 1.0,
    "tissue_lowres_scalef": 0.1875,
    "fiducial_diameter_fullres": 70.5843019138272,
    "tissue_hires_scalef": 1.0
}
Is there a specific email I can contact to give access to the files?

**Comment by ktpolanski** [comment]
Thanks for the information, you somehow seem to have lost one of the maximum values but I have enough to work with.

There is no bug in bin2cell, it looks like the specified full resolution morphology image was not actually used for Spaceranger. As you can see, the values of the spatial coordinates (as loaded from the Spaceranger output, and not modified in any form by bin2cell itself) are very small, not matching the 30k+ pixels you see in the full resolution image. Additionally, the `tissue_hires_scalef` is 1. This value is meant to be taken and used to multiply the full morphology image spatial coordinates to get them shrunken down to display on the "high resolution" image Spaceranger makes, which has a maximum dimension of 6000 pixels. Dividing 6000 by the scale factor gets the dimension of the original morphology image. 6000 / 1 is, well, 6000. Not the 35k+ from the display you showed earlier.

Just as an example, the mouse brain demo data has a `tissue_hires_scalef` of 0.2505533 and the full resolution morphology image has a maximum dimension of 23947.

As is, I don't think there's much point to trouble you with a file upload. Rerun Spaceranger with the actual full resolution image.

**Comment by ktpolanski** [comment]
This exchange inspired the creation of `b2c.actual_vs_inferred_image_shape()` to judge the compatibility of image shape rather than checking for bin overlap. This catches corner cases like this one, though the vast majority of mismatches will still probably load an image that's too small. Implemented in 0.3.2.
